processor/memorylimiterprocessor/memorylimiter_test.go

```diff
@@ -440,3 +108,379 @@ func TestNoDataLoss(t *testing.T) {
+			memAlloc:    800,
+			expectError: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			memorylimiter.GetMemoryFn = totalMemory
+			memorylimiter.ReadMemStatsFn = func(ms *runtime.MemStats) {
+				ms.Alloc = tt.memAlloc + tt.ballastSize
+			}
+
+			ml, err := newMemoryLimiterProcessor(processortest.NewNopCreateSettings(), tt.mlCfg)
+			require.NoError(t, err)
+			mp, err := processorhelper.NewMetricsProcessor(
+				context.Background(),
+				processortest.NewNopCreateSettings(),
+				tt.mlCfg,
+				consumertest.NewNop(),
+				ml.processMetrics,
+				processorhelper.WithCapabilities(processorCapabilities),
+				processorhelper.WithStart(ml.start),
+				processorhelper.WithShutdown(ml.shutdown))
+			require.NoError(t, err)
+
+			assert.NoError(t, mp.Start(ctx, &host{ballastSize: tt.ballastSize}))
+			ml.memlimiter.CheckMemLimits()
+			err = mp.ConsumeMetrics(ctx, md)
+			if tt.expectError {
+				assert.Equal(t, memorylimiter.ErrDataRefused, err)
+			} else {
+				assert.NoError(t, err)
+			}
+			assert.NoError(t, mp.Shutdown(ctx))
+		})
+	}
+	t.Cleanup(func() {
+		memorylimiter.GetMemoryFn = iruntime.TotalMemory
+		memorylimiter.ReadMemStatsFn = runtime.ReadMemStats
+	})
+}
+
+// TestTraceMemoryPressureResponse manipulates results from querying memory and
+// check expected side effects.
+func TestTraceMemoryPressureResponse(t *testing.T) {
+	td := ptrace.NewTraces()
+	ctx := context.Background()
+
+	tests := []struct {
+		name        string
+		mlCfg       *Config
+		ballastSize uint64
+		memAlloc    uint64
+		expectError bool
+	}{
+		{
+			name: "Below memAllocLimit",
+			mlCfg: &Config{
+				CheckInterval:         time.Second,
+				MemoryLimitPercentage: 50,
+				MemorySpikePercentage: 1,
+			},
+			ballastSize: 0,
+			memAlloc:    800,
+			expectError: false,
+		},
+		{
+			name: "Above memAllocLimit",
+			mlCfg: &Config{
+				CheckInterval:         time.Second,
+				MemoryLimitPercentage: 50,
+				MemorySpikePercentage: 1,
+			},
+			ballastSize: 0,
+			memAlloc:    1800,
+			expectError: true,
+		},
+		{
+			name: "Below memAllocLimit accounting for ballast",
+			mlCfg: &Config{
+				CheckInterval:         time.Second,
+				MemoryLimitPercentage: 50,
+				MemorySpikePercentage: 1,
+			},
+			ballastSize: 1000,
+			memAlloc:    800,
+			expectError: false,
+		},
+		{
+			name: "Above memAllocLimit even accounting for ballast",
+			mlCfg: &Config{
+				CheckInterval:         time.Second,
+				MemoryLimitPercentage: 50,
+				MemorySpikePercentage: 1,
+			},
+			ballastSize: 1000,
+			memAlloc:    1800,
+			expectError: true,
+		},
+		{
+			name: "Below memSpikeLimit",
+			mlCfg: &Config{
+				CheckInterval:         time.Second,
+				MemoryLimitPercentage: 50,
+				MemorySpikePercentage: 10,
+			},
+			ballastSize: 0,
+			memAlloc:    800,
+			expectError: false,
+		},
+		{
+			name: "Above memSpikeLimit",
+			mlCfg: &Config{
+				CheckInterval:         time.Second,
+				MemoryLimitPercentage: 50,
```

Comment:
Why `MemoryLimitPercentage` instead of `MemoryLimitMiB`? We should not rely on the test environment

Reply:
since the `MemoryLimitMiB`, changes the int is in mb unit, the resulting value would be large. too large to easily understand in the test

```
memAllocLimit := uint64(cfg.MemoryLimitMiB) * mibBytes
memSpikeLimit := uint64(cfg.MemorySpikeLimitMiB) * mibBytes
```
(how the original is)

the original test modified the `memUsageChecker` values directly so it could achieve small values for tests. The percentage would allow the tests values to continue be small values

Reply:
Ok, sounds good.